### PR TITLE
[1.21] April 4 CVEs

### PR DESCRIPTION
**Pull request by jbohanon**
Backports of April 4 CVE fixes from upstream's release/v1.22 ([commits page](https://github.com/envoyproxy/envoy/commits/release/v1.22)). This is an attempt to capture relevant pieces of the diff from before Apr 4 to the release v1.22.10 ([diff](https://github.com/envoyproxy/envoy/compare/cd78a8408333b97d9e9ff7979e7f8c1508626e26...v1.22.10)).
- 80d928f20141f0cbd948a37ac7c6364044dcabac
- ec042c15cf331157a51b15c8fe1af16fc14f2611
- 4c50612c421ef52228232eba3d332d6c2c4cba9b
- 5ced3041fc2da22fcf72fd621ce432aa26024caa
- 6546a08e92edb662658f1d78b8f6c1b8ed53ac8d
- 06ff1dd4db0fc8b2d00bf1cb718948fcba135c45
- 91e9cb0ae869244362b6f973c5b7beb56a6b74b5

Backporting from release/v1.22 was selected to attempt to minimize merge conflicts from unrelated features/fixes that have entered the codebase in later versions. 

These changes are being ported onto v1.21.5 due to some breaking changes introduced causing difficulties compiling, specifically in [this PR](https://github.com/envoyproxy/envoy/pull/19275). These will need to be resolved if we ever need to bump to a later released version of Envoy 1.21.x, but for now this will unblock work requiring a patched `envoy-gloo` release.

I have manually gone through each of the relevant commits and validated that necessary logic and testing of change behavior was included.

### CODE REVIEW

File: source/common/quic/envoy_quic_server_stream.cc [hunk 1]
```diff
@@ -158,9 +159,12 @@ void EnvoyQuicServerStream::OnInitialHeadersComplete(bool fin, size_t frame_len,
...
     return;
   }
-  if (Http::HeaderUtility::requestHeadersValid(*headers) != absl::nullopt ||
-      Http::HeaderUtility::checkRequiredRequestHeaders(*headers) != Http::okStatus() ||
+
+  if (Http::HeaderUtility::checkRequiredRequestHeaders(*headers) != Http::okStatus() ||
+      Http::HeaderUtility::checkValidRequestHeaders(*headers) != Http::okStatus() ||
       (headers->Protocol() && !http3_options_.allow_extended_connect())) {
+      // Note: Cherry-pick included below line, but it looks like v1.21 did not yet incorporate spdy here
+      /* (headers->Protocol() && !spdy_session()->allow_extended_connect())) { */
```
Comment (jbohanon) [inline source/common/quic/envoy_quic_server_stream.cc]: ec042c15cf331157a51b15c8fe1af16fc14f2611
ok to remove

File: source/common/runtime/runtime_features.cc [hunk 2]
```diff
@@ -91,6 +95,8 @@ constexpr const char* runtime_features[] = {
     "envoy.reloadable_features.use_dns_ttl",
     "envoy.reloadable_features.use_observable_cluster_name",
     "envoy.reloadable_features.validate_connect",
+    // DO NOT MERGE until validating this feature exists
```
Comment (jbohanon) [inline source/common/runtime/runtime_features.cc]: ec042c15cf331157a51b15c8fe1af16fc14f2611
ok

File: test/common/http/header_utility_test.cc [hunk 3]
```diff
@@ -939,5 +939,84 @@ TEST(ValidateHeaders, ContentLength) {
   EXPECT_TRUE(should_close_connection);
 }
 
+// NET-NEW 1.22
+// These tests were added in the changes being backported from 1.22
```
Comment (jbohanon) [inline test/common/http/header_utility_test.cc]: ec042c15cf331157a51b15c8fe1af16fc14f2611
Make best-effort attempt to make this test compile

File: test/common/http/header_utility_test.cc [hunk 4]
```diff
@@ -939,5 +939,84 @@ TEST(ValidateHeaders, ContentLength) {
...
+//                                     {":authority", "foo.com:80"},
+//                                     {"x-foo", "hello\r\n\r\nGET /evil HTTP/1.1"}};
+//    EXPECT_NE(Http::okStatus(), HeaderUtility::checkValidRequestHeaders(headers));
+//  }
+//}
+//#endif
+// END NET-NEW v1.22
+
+// EXTANT IN 1.22
+// These tests existed prior to the changes being backported from 1.22
```
Comment (jbohanon) [inline test/common/http/header_utility_test.cc]: ec042c15cf331157a51b15c8fe1af16fc14f2611
Ok to leave out these tests, unless they come for free when fixing `ForbiddenCharacters`

File: source/common/runtime/runtime_features.cc [hunk 5]
```diff
@@ -82,7 +82,11 @@ constexpr const char* runtime_features[] = {
     "envoy.reloadable_features.proxy_102_103",
     "envoy.reloadable_features.remove_legacy_json",
     "envoy.reloadable_features.require_strict_1xx_and_204_response_headers",
+    // DO NOT MERGE until validating this feature exists
```
Comment (jbohanon) [inline source/common/runtime/runtime_features.cc]: 91e9cb0ae869244362b6f973c5b7beb56a6b74b5
ok

File: source/common/runtime/runtime_features.cc [hunk 6]
```diff
@@ -82,7 +82,11 @@ constexpr const char* runtime_features[] = {
...
+    "envoy.reloadable_features.sanitize_original_path",
     "envoy.reloadable_features.send_strict_1xx_and_204_response_headers",
+    // DO NOT MERGE until validating this feature exists
```
Comment (jbohanon) [inline source/common/runtime/runtime_features.cc]: 5ced3041fc2da22fcf72fd621ce432aa26024caa ok

File: test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc [hunk 7]
```diff
@@ -571,26 +662,60 @@ TEST_P(ExtProcIntegrationTest, GetAndSetHeadersAndTrailersOnResponse) {
...
+      });
 
   verifyDownstreamResponse(*response, 200);
   ASSERT_TRUE(response->trailers());
   EXPECT_THAT(*(response->trailers()), SingleHeaderValueIs("x-test-trailers", "Yes"));
   EXPECT_THAT(*(response->trailers()), SingleHeaderValueIs("x-modified-trailers", "xxx"));
 }
 
+// EXTANT IN 1.22
+// This test existed prior to the changes being backported from 1.22
```
Comment (jbohanon) [inline test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc]: 5ced3041fc2da22fcf72fd621ce432aa26024caa
signature change only; doesn't seem needed. will make reasonable effort to compile regardless

File: test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc [hunk 8]
```diff
@@ -931,6 +1087,49 @@ TEST_P(ExtProcIntegrationTest, GetAndRespondImmediatelyWithBadStatus) {
   EXPECT_EQ("{\"reason\": \"Because\"}", response->body());
 }
 
+// EXTANT IN 1.22
+// These tests existed prior to the changes being backported from 1.22
```
Comment (jbohanon) [inline test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc]: 5ced3041fc2da22fcf72fd621ce432aa26024caa
signature change only; doesn't seem needed. will make reasonable effort to compile regardless

**Review by ben-taussig-solo — COMMENTED** [review]
> These changes are being ported onto v1.21.5 due to some breaking changes introduced causing difficulties compiling, specifically in https://github.com/envoyproxy/envoy/pull/19275.

I guess my biggest concern is whether envoy-gloo/envoy-gloo-ee depend on any of the changes that went into 1.21.6 (as far as I can tell the only upstream v1.21.x release published after v1.21.5)

Otherwise, this seems good to me!

**Comment by jbohanon** [comment]
> > These changes are being ported onto v1.21.5 due to some breaking changes introduced causing difficulties compiling, specifically in [envoyproxy#19275](https://github.com/envoyproxy/envoy/pull/19275).
> 
> I guess my biggest concern is whether envoy-gloo/envoy-gloo-ee depend on any of the changes that went into 1.21.6 (as far as I can tell the only upstream v1.21.x release published after v1.21.5)
> 
> Otherwise, this seems good to me!

The commit on which `envoy-gloo` release-v1.21 currently depends (l[ink](https://github.com/solo-io/envoy-gloo/blob/release-v1.21/bazel/repository_locations.bzl)) is the commit sha for release v1.21.5 ([link](https://github.com/envoyproxy/envoy/releases/tag/v1.21.5)), and `envoy-gloo-ee` consumes upstream transitively through `envoy-gloo`

**Comment by nfuden** [comment]
i wonder if we should just call this release-1.21 as that branch will easily get fastforwarded by upstream commits

**Comment by jbohanon** [comment]
> i wonder if we should just call this release-1.21 as that branch will easily get fastforwarded by upstream commits

We're explicitly _not_ following upstream commits since we are branched at `1.21.5` and not tracking upstream `release-v1.21`

**Review by nfuden — APPROVED** [review]

**Comment by jbohanon** [comment]
Proceeding with merge despite failing checks from upstream since our consuming repo is passing CI with these changes, indicating successful builds and tests ([envoy-gloo PR](https://github.com/solo-io/envoy-gloo/pull/225))

**Pull request merged**